src/backend/distributed/utils/resource_lock.c

```diff
@@ -1060,3 +1061,143 @@ CitusLockTableAclCheck(Oid relationId, LOCKMODE lockmode, Oid userId)
+ * to the metadata.
+ */
+void
+AcquireDistributedLockOnRelations(List *relationIdList, LOCKMODE lockMode)
+{
+	if (!EnableMetadataSync)
+	{
+		return;
+	}
+
+	const char *lockModeText = LockModeToLockModeText(lockMode);
+
+	UseCoordinatedTransaction();
+
+	StringInfo lockRelationsCommand = makeStringInfo();
+	appendStringInfo(lockRelationsCommand, "%s;\n", DISABLE_DDL_PROPAGATION);
+
+	List *relationIdsToLock = NIL;
+	bool partialLockStmt = false;
+
+	Oid relationId = InvalidOid;
+	foreach_oid(relationId, relationIdList)
+	{
+		EnsureCanAcquireLock(relationId, lockMode);
+
+		CitusTableCacheEntry *tableCacheEntry = GetCitusTableCacheEntry(relationId);
+
+		/* skip citus tables which are not seen by workers */
+		if (tableCacheEntry->partitionMethod == DISTRIBUTE_BY_RANGE ||
+			tableCacheEntry->partitionMethod == DISTRIBUTE_BY_APPEND)
+		{
+			continue;
+		}
+
+		char *qualifiedRelationName = generate_qualified_relation_name(relationId);
+
+		/*
+		 * As of pg14 we cannot use LOCK to lock a FOREIGN TABLE
+		 * so we have to use the lock_relation_if_exist udf
+		 */
+		if (get_rel_relkind(relationId) == RELKIND_FOREIGN_TABLE)
+		{
+			/* finish the partial lock statement */
+			if (partialLockStmt)
+			{
+				appendStringInfo(lockRelationsCommand, " IN %s MODE;\n", lockModeText);
+				partialLockStmt = false;
+			}
+
+			/* use lock_relation_if_exits to lock foreign table */
+			appendStringInfo(lockRelationsCommand, LOCK_RELATION_IF_EXISTS,
+							 quote_literal_cstr(qualifiedRelationName),
+							 quote_literal_cstr(lockModeText));
+			appendStringInfoChar(lockRelationsCommand, '\n');
+		}
+		else if (partialLockStmt)
+		{
+			/* append relation to partial lock statement */
+			appendStringInfo(lockRelationsCommand, ", %s", qualifiedRelationName);
+		}
+		else
+		{
+			/* start a partial lock statement */
+			appendStringInfo(lockRelationsCommand, "LOCK %s", qualifiedRelationName);
+			partialLockStmt = true;
+		}
+
+		relationIdsToLock = lappend_oid(relationIdsToLock, relationId);
+	}
+
+	if (list_length(relationIdsToLock) == 0)
+	{
+		return;
+	}
+
+	if (partialLockStmt)
+	{
+		appendStringInfo(lockRelationsCommand, " IN %s MODE;\n", lockModeText);
+	}
+
+	appendStringInfo(lockRelationsCommand, ENABLE_DDL_PROPAGATION);
+	const char *lockCommand = lockRelationsCommand->data;
+
+
+	List *workerNodeList = TargetWorkerSetNodeList(METADATA_NODES, NoLock);
+
+	/*
+	 * We want to acquire locks in the same order across the nodes.
+	 * Although relation ids may change, their ordering will not.
+	 */
+	workerNodeList = SortList(workerNodeList, CompareWorkerNodes);
+
+	int32 localGroupId = GetLocalGroupId();
+
+	WorkerNode *workerNode = NULL;
+	foreach_ptr(workerNode, workerNodeList)
+	{
+		const char *nodeName = workerNode->workerName;
+		int nodePort = workerNode->workerPort;
+
+		/* if local node is one of the targets, acquire the lock locally */
+		if (workerNode->groupId == localGroupId)
```

Comment:
I believe not in this PR, but for the LOCK TABLE pr, we'd need to imitate (or directly call [LockTableCommand](https://github.com/postgres/postgres/blob/master/src/backend/commands/lockcmds.c#L42-L64) for local node, because when the table is a view or a partitioned table, it needs to recursively find all relevant tables.

We already have `lockCommand`, so we should perhaps simply do `ExecuteUtilityCommand()`, where we already have `enable/disable dll` checks.

I'm fine to do it in this PR as well

Reply:
My impression was that the standard utility is going to do that for us, but in order to enforce the same locking order throughout, you are right in that it is probably best to do it here 👍

Reply:
yeap, we'll be re-acquiring the same locks during standard utility, but here we are ensuring the worker-node order.